src/gateway/WebSocketClient.ts

```diff
@@ -0,0 +1,127 @@
+import WebSocket from 'ws';
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import { EVENTS, Payload } from './constants';
+import { GATEWAY_OPCODES, GATEWAY_CLOSE_EVENT_CODES } from '../util/Constants';
+
+export default class WebSocketClient {
+  private _intents: number;
+  private _token: string;
+  private _url: string;
+  private heartBeatInterval!: NodeJS.Timeout;
+  private lastHeartbeatAck = true;
+  private seq!: number;
+  private sessionID!: string;
+  private ws!: WebSocket;
+
+  constructor(token: string, wsURL: string, intents: number) {
+    this._token = token;
+    this._url = wsURL;
+    this._intents = intents;
+
+    this._identify = this._identify.bind(this);
+    this._onClose = this._onClose.bind(this);
+    this._onMessage = this._onMessage.bind(this);
+  }
+
+  connect() {
+    this.ws = new WebSocket(this._url);
+    this.ws.on('open', this._identify);
+    this.ws.on('close', this._onClose);
+    this.ws.on('message', this._onMessage);
+  }
+
+  send(op: GATEWAY_OPCODES, data: any) {
+    if (this.ws.readyState === WebSocket.OPEN) {
+      this.ws.send(JSON.stringify({
+        op,
+        d: data,
+      }));
+    }
+  }
+
+  private _heartbeat(expected: boolean) {
+    if (expected) {
+      if (!this.lastHeartbeatAck) {
+        this._terminate();
+        this.connect();
+        this._resume();
+        this.lastHeartbeatAck = true;
+      } else {
+        this.lastHeartbeatAck = false;
+      }
+    }
+    this.send(GATEWAY_OPCODES.HEARTBEAT, this.seq || null);
+  }
+
+  private _identify() {
+    this.send(GATEWAY_OPCODES.IDENTIFY, {
+      token: this._token,
+      intents: this._intents,
+      properties: {
+        $os: process.platform,
+        $browser: 'Dis.ts',
+        $device: 'Dis.ts',
+      },
```

Comment:
At some point we might want to allow people to input whatever they want here.. that's for a later time though